### PR TITLE
[FLINK-6237] [table] support RAND and RAND_INTEGER on SQL

### PR DESCRIPTION
**Pull request by godfreyhe**
support RAND and RAND_INTEGER with and without seed on SQL.
like: 
RAND([seed]), 
RAND_INTEGER([seed, ] bound)

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -1494,6 +1519,22 @@ class CodeGenerator(
...
+    * Adds a reusable [[java.util.Random]] to the member area of the generated [[Function]].
+    *
+    * @return member variable term
+    */
+  def addReusableRandom(): String = {
+    val fieldTerm = newName("random")
+
+    val field =
+      s"""
+         |transient java.util.Random $fieldTerm = null;
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: `Random` implements `Serializable`. I think we initialize `Random` already here, remove `transient` and add `final`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 2]
```diff
@@ -911,6 +911,64 @@ object ScalarOperators {
...
+  def generateRand(
+    randField: String,
+    seedExpr: GeneratedExpression,
+    resultType: TypeInformation[_])
+  : GeneratedExpression = {
+    val resultTerm = newName("result")
+    val resultTypeTerm = primitiveTypeTermForTypeInfo(resultType)
+    val randCode = if (seedExpr != null) {
+      s"""
+         |if ($randField == null) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: If we initialize `Random` before, we do not need to this condition.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 3]
```diff
@@ -911,6 +911,64 @@ object ScalarOperators {
...
+    seedExpr: GeneratedExpression,
+    boundExpr: GeneratedExpression,
+    resultType: TypeInformation[_])
+  : GeneratedExpression = {
+    assert(boundExpr != null)
+    val resultTerm = newName("result")
+    val resultTypeTerm = primitiveTypeTermForTypeInfo(resultType)
+    val randCode = if (seedExpr != null) {
+      s"""
+         |if ($randField == null) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: If we initialize `Random` before, we do not need this condition.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala [hunk 4]
```diff
@@ -168,6 +168,18 @@ class SqlExpressionTest extends ExpressionTestBase {
     testSqlApi("ELEMENT(ARRAY['HELLO WORLD'])", "HELLO WORLD")
   }
 
+  @Test
+  def testRand(): Unit = {
+    val random = new java.util.Random(1)
+    testSqlApi("RAND(1)", random.nextDouble().toString)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala]: Check a second time that the next random number is different.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala [hunk 5]
```diff
@@ -168,6 +168,18 @@ class SqlExpressionTest extends ExpressionTestBase {
...
+  }
+
+  @Test
+  def testRandInteger(): Unit = {
+    val random = new java.util.Random(1)
+    testSqlApi("RAND_INTEGER(1, 10)", random.nextInt(10).toString)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala]: Check a second time.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 6]
```diff
@@ -911,6 +911,64 @@ object ScalarOperators {
     }
   }
 
+  def generateRand(
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: The ScalarOperators class is intended for basic operations such as plus, minus, equals etc. Can you add the functions to `FunctionGenerator` class?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala [hunk 7]
```diff
@@ -1106,6 +1106,31 @@ class CodeGenerator(
         requireArray(array)
         generateArrayElement(this, array)
 
+      // RAND([seed])
+      case RAND if isNumeric(resultType) && operands.size <= 1 =>
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala]: We don't need this changes if the method is implemented using a `CallGenerator` in `FunctionGenerator`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala [hunk 8]
```diff
@@ -911,6 +911,64 @@ object ScalarOperators {
...
+    seedExpr: GeneratedExpression,
+    resultType: TypeInformation[_])
+  : GeneratedExpression = {
+    val resultTerm = newName("result")
+    val resultTypeTerm = primitiveTypeTermForTypeInfo(resultType)
+    val randCode = if (seedExpr != null) {
+      s"""
+         |if ($randField == null) {
+         |  ${seedExpr.code}
+         |  $randField = new java.util.Random(${seedExpr.resultTerm});
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala]: What happens if `seedExpr.nullTerm` is true? We need a case distinction for both null check enabled and disabled.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala (hunk 4 above)
Comment (godfreyhe) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala]: This test case can only test a expression on one row. So I will add a IT case to test rand and rand_integer on multi-rows.

**Comment by godfreyhe** [comment]
thanks for your advice @fhueske @twalthr.  I had update the PR.
I will implement rand and rand_integer functions for the Scala and Java Table API next immediately, or should I submit a new PR ?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala (hunk 4 above)
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/SqlExpressionTest.scala]: I see, I think in that case it should be OK, to keep the test as it is.

**Comment by fhueske** [comment]
Thanks for the update @godfreyhe! 
Looks good, IMO.

I think it would be good to add the Table API functions and documentation with this PR.

Thanks, Fabian

**Comment by twalthr** [comment]
Thanks @godfreyhe. I will merge this.

**Comment by twalthr** [comment]
Unfortunately, I found two issues that need to be fixed before we can merge this. According to SQL the seed and bound can also be defined by an other column. Right now `rand(f7)` leads to a compiler exception. We should check if the arguments are constant and only then move it to the reusable area.

**Comment by godfreyhe** [comment]
@twalthr do you mean that sql like ```select a, rand(a) from table``` does not work?

**Comment by twalthr** [comment]
@godfreyhe exactly. We should check if the argument is literal or not.

**Comment by godfreyhe** [comment]
@twalthr, I had update this PR.

**Comment by twalthr** [comment]
@godfreyhe sorry for the delay. I will rebase this PR and merge it.

**Pull request closed without merging**